### PR TITLE
Refactor API Gateway

### PR DESCRIPTION
**Pull request by Eelviny**
### Step 1: Link to Jira issue
https://stackstate.atlassian.net/browse/STAC-13306

### Step 2: Description of changes
- Refactor API Gateway
- Add IAM permission checks
- Add Schematics
- Fix API response tests to show needed data
- Add API gateway to CloudFormation integration tests

### Step 3: Did you add / update tests for your changes in the right area?
- [x] Unit/Component test
- [x] Integration test	

### Step 4: I'm confident that everything is properly tested:
I got a PO / QA Approval by:
- [ ] Name

### Step 5: Can we ship this feature to production?
- [ ] Yes, I'm proud of my work. Ship it! :ship:

### CODE REVIEW

File: aws_topology/tests/json/apigateway/get_stages_074f0b3.json [hunk 1]
```diff
@@ -12,12 +12,9 @@
...
             "lastUpdatedDate": 1502091405,
-            "tags": [
-                {
-                    "key": "StageTagKey1",
-                    "value": "StageTagValue1"
-                }
-            ]
+            "tags": {
+                "StageTagKey1": "StageTagValue1"
+            }
```
Comment (Eelviny) [inline aws_topology/tests/json/apigateway/get_stages_074f0b3.json:17]: This needs integration testing - API response is different to what is in boto3 docs

File: aws_topology/stackstate_checks/aws_topology/resources/api_gateway.py [hunk 2]
```diff
@@ -25,135 +35,206 @@ def create_method_arn(region=None, account_id=None, resource_id=None, **kwargs):
...
-                                )
-                                self.emit_relation(method_arn, service_integration_urn, "uses service", {})
+                )
+            )
+            self.emit_component(resource_arn, "aws.apigateway.resource", resource_data)
+            self.emit_relation(stage_arn, resource_arn, "uses service", {})
+
+            for method_id in resource.resourceMethods.keys():
+                self.process_resource_method(
+                    data["resourceMethods"][method_id],
```
Comment (Jacco) [inline aws_topology/stackstate_checks/aws_topology/resources/api_gateway.py:189]: Use the schematic here

**Review by Jacco — CHANGES_REQUESTED** [review]

File: aws_topology/stackstate_checks/aws_topology/resources/api_gateway.py (hunk 2 above)
Comment (Eelviny) [reply aws_topology/stackstate_checks/aws_topology/resources/api_gateway.py:189]: I need the original json object, schematics would lose all values that I haven't defined. Line 190 includes the schematics version

File: aws_topology/tests/json/apigateway/get_stages_074f0b3.json (hunk 1 above)
Comment (Eelviny) [reply aws_topology/tests/json/apigateway/get_stages_074f0b3.json:17]: API response was wrong, leaving it like this

**Review by Jacco — APPROVED** [review]

File: aws_topology/stackstate_checks/aws_topology/resources/api_gateway.py [hunk 3]
```diff
@@ -25,135 +35,218 @@ def create_method_arn(region=None, account_id=None, resource_id=None, **kwargs):
...
+    @set_required_access_v2("apigateway:GET")
+    def collect_methods(self, resource_data, rest_api_id):
+        for method in resource_data.get("resourceMethods", {}).keys():
+            # The boto3 docs suggest that methods are returned in the get_resources call; this is a lie.
+            if not resource_data["resourceMethods"][method]:
+                resource_data["resourceMethods"][method] = (
+                    self.client.get_method(restApiId=rest_api_id, resourceId=resource_data.get("id"), httpMethod=method)
+                    or {}
+                )
+        print(resource_data)
```
Comment (Eelviny) [inline aws_topology/stackstate_checks/aws_topology/resources/api_gateway.py:83]: todo remove

**Review by vishalxebia — APPROVED** [review]

**Pull request merged**